### PR TITLE
Free hosting trial: Add CTA to the Business plan card

### PR DESCRIPTION
**Pull request by zaguiini**
Related to p1696387524534499-slack-C041RHH38NQ.

## Proposed Changes

Displays the "Try for free" CTA alongside the "Get Business" button inside the plan card:

| Pricing grid | Plan comparison grid |
| ------------ | --------------------- |
| image | image |

I tried to devise an implementation as agnostic as possible so that we could add more trial CTAs (e.g., Commerce) to the page later (see p1696516487969099/1696387524.534499-slack-C041RHH38NQ.)

## Testing Instructions

1. Enable the `plans/hosting-trial` feature flag
2. Enter testing mode in the store admin
3. Navigate to `/setup/new-hosted-site`
4. Fill the site options step input and click continue
5. Verify that the "Try for free" CTA shows up alongside the "Get Business" CTA, both in the pricing grid and the plan comparison grid
6. Verify that as you scroll the page the trial CTA gets stuck to the page as well
7. Click the "try for free" CTA and verify that a new trial site gets created
8. Retry the flow but click the Business CTA, verify you're sent to checkout normally

### CODE REVIEW

File: client/my-sites/plans-grid/components/actions.tsx
Comment (chriskmnds) [inline client/my-sites/plans-grid/components/actions.tsx]: Let's wrap this with useCallback for reference integrity

File: client/my-sites/plans-grid/hooks/use-get-free-trial-slug-for-plan.ts
Comment (chriskmnds) [inline client/my-sites/plans-grid/hooks/use-get-free-trial-slug-for-plan.ts]: Cannot have Calypso state references in plans-grid folder. Think of this as a package already.

**Review by chriskmnds — DISMISSED** [review]
Thanks for the code. I need to take a closer look to ensure this work is aligned with roadmap, as it touches several files we are closely monitoring. 

Left a couple of obvious points that will need to be resolved for now. I can take a closer look on Monday. Also cc @Automattic/martech-decepticons

File: client/my-sites/plans-grid/hooks/use-get-free-trial-slug-for-plan.ts
Comment (zaguiini) [reply client/my-sites/plans-grid/hooks/use-get-free-trial-slug-for-plan.ts]: It's outside the `npm-ready` folder though, and I see we already have other hooks in that folder with Calypso references.

Comment (chriskmnds) [reply client/my-sites/plans-grid/hooks/use-get-free-trial-slug-for-plan.ts]: Yeah, we don't want to introduce new ones. Sorry if existence of folder directed towards thinking that it's ok to do so outside. Folder only there for maintenance/organisational reasons to help with ongoing refactors.

see "Plans code to NPM" Project Thread on Martech P2 for more dstails

Comment (zaguiini) [reply client/my-sites/plans-grid/hooks/use-get-free-trial-slug-for-plan.ts]: Oh, alright. I'll change it 👍

Comment (zaguiini) [reply client/my-sites/plans-grid/hooks/use-get-free-trial-slug-for-plan.ts]: Done @chriskmnds please leave another round of review.

File: client/my-sites/plans-grid/components/actions.tsx
Comment (chriskmnds) [inline client/my-sites/plans-grid/components/actions.tsx]: Let's import from `@wordpress/element` instead

File: client/my-sites/plans-grid/components/actions.tsx
Comment (chriskmnds) [inline client/my-sites/plans-grid/components/actions.tsx]: Do we need a separate handler for this? I'd think `handleUpgradeButtonClick` should suffice for both cases.

File: client/my-sites/plans-grid/index.tsx
Comment (chriskmnds) [inline client/my-sites/plans-grid/index.tsx]: I think this adds unnecessary clutter, depends on `flowName` that we want to steer away from (we are hoping to completely uncouple grid from this concept), and also breaks away from the data foundation used -> all of the relevant metadata for grid plans is provided through the `GridPlan` type.

Have we considered feeding the free trial slugs through [use-grid-plans](https://github.com/Automattic/wp-calypso/blob/trunk/client/my-sites/plans-grid/hooks/npm-ready/data-store/use-grid-plans.tsx) hook? So I would have expected a hook to be called from there at the top level, then grab the available trial slugs and attach to the grid plans/objects returned, just like we do for the other metadata. 

maybe take a look at how `usePricingMetaForGridPlans` or `usePricedAPIPlans` are used and follow along?

but was there some other reason we chose to propagate this callback in the components instead? 🤔

**Review by chriskmnds — DISMISSED** [review]
Thanks for following up :-) 
I left a few more comments below.

File: client/my-sites/plans-grid/index.tsx
Comment (chriskmnds) [reply client/my-sites/plans-grid/index.tsx]: > maybe take a look at how `usePricingMetaForGridPlans` or `usePricedAPIPlans` are used and follow along?

and also try and use `intent` (in combination with `isInSignup` if you want to target just signup) instead of `flowName`.

Comment (zaguiini) [reply client/my-sites/plans-grid/index.tsx]: >but was there some other reason we chose to propagate this callback in the components instead? 🤔

Sheer ignorance 😁 Thanks for the help, I'll try and rewrite the code!

File: packages/calypso-products/src/constants/types.ts
Comment (zaguiini) [inline packages/calypso-products/src/constants/types.ts:46]: We did not have these Jetpack plan types added to the list for some reason. I went ahead and added them.

File: packages/calypso-products/src/types.ts
Comment (zaguiini) [inline packages/calypso-products/src/types.ts:230]: I changed it to a union type of the `TYPES_LIST`. Nothing broke, and we have more type safety now. I wonder why it wasn't like that before?

**Review by zaguiini — COMMENTED** [review]
@chriskmnds I think this is much better now ✨ Thanks for all the help 🙇

File: client/my-sites/plans-features-main/index.tsx
Comment (chriskmnds) [inline client/my-sites/plans-features-main/index.tsx]: We can/should turn this into a hook and call directly from `useGridPlans`. that's been the case with everything else (hooks > props). We'll be porting and organising all of these data hooks into `@automattic/data-stores` down the line, with a simple interface for the consumer end. The criteria right now are more or less:

1. If no Calypso dependencies, then hook can live under `/plans-grid/hooks/npm-ready` and called directly
2. Otherwise, under `/plans-features-main/hooks` and pass through (the hook) as a prop to `use-grid-plans`

In this case, you could just be passing `eligibleForFreeHostingTrial` to use-grid-plans.

Ok for this to happen in a follow-up (I can also do this)

File: client/my-sites/plans-grid/components/comparison-grid/index.tsx
Comment (chriskmnds) [inline client/my-sites/plans-grid/components/comparison-grid/index.tsx]: I know you are probably following along to how `planSlug` is passed, but passing `freeTrialPlanSlug` around is unnecessary. I think you can remove it entirely and retrieve it from the plans in context e.g. see: https://github.com/Automattic/wp-calypso/blob/trunk/client/my-sites/plans-grid/components/actions.tsx#L388-L392

(we may revise the use of context down the line, but it will likely be a replacement with a data-store, not component props)

File: client/my-sites/plans-grid/components/features-grid/index.tsx
Comment (chriskmnds) [inline client/my-sites/plans-grid/components/features-grid/index.tsx:401]: We can wrap this with useCallback to pass same reference on rerenders. I know it wasn't done before, so totally fine to ignore.

File: client/my-sites/plans-grid/components/comparison-grid/index.tsx
Comment (chriskmnds) [inline client/my-sites/plans-grid/components/comparison-grid/index.tsx:462]: Similar here. We can wrap this with `useCallback` to pass same reference on rerenders. I know it wasn't done before, so totally fine to ignore.

File: client/my-sites/plans-grid/components/actions.tsx
Comment (chriskmnds) [inline client/my-sites/plans-grid/components/actions.tsx]: Similar here. Let's wrap this with `useCallback` to force a single reference across rerenders. and also for the case below

Of course, let me know if you think otherwise. :-)

File: packages/calypso-products/src/types.ts
Comment (chriskmnds) [reply packages/calypso-products/src/types.ts:230]: Nice. 

> I wonder why it wasn't like that before?

No idea. A lot of things weren't done from before :D

**Review by chriskmnds — COMMENTED** [review]
Thanks for the updates. I left a few more comments, mostly optimizations (except for the case where we pass the prop around for no need). Otherwise tested and looks good (but run a few more tests pls as I ran out of time)

One other thing, I notice the plans shown under `/plans/[ new-hosted-site ]` (e.g. `plans/asdf3232.wpcomstaging.com?flags=plans/hosting-trial`) shows the default list of plans:

Screenshot 2023-10-09 at 3 25 30 PM

You can make the same changes we applied for `/setup/newsletter` and only show the same plans shown on `/setup/new-hosted-site/plans` in admin. (in case that's something of interest)

**Comment by mpkelly** [comment]
I tested this and it looks good. I created a site with a trail plan using the new button. I checked `/start/plans`, and there is no change as expected; same when going through the `setup/import-hosted-site`.

File: client/my-sites/plans-grid/components/features-grid/index.tsx
Comment (zaguiini) [reply client/my-sites/plans-grid/components/features-grid/index.tsx:401]: I can't do this here, the plan slug from the mapped data is used as a dependency (not directly accessible through `this.props`), and this is not a functional component.

File: client/my-sites/plans-grid/components/actions.tsx
Comment (zaguiini) [reply client/my-sites/plans-grid/components/actions.tsx]: Eh, I do have my reservations with useCallback ([see this post](https://kentcdodds.com/blog/usememo-and-usecallback#why-is-usecallback-worse)).

The gains are marginal and we're introducing the overhead of calling `useCallback` twice as we have two different click responses there so it's impossible _not_ to define a new arrow function unless we use callback twice... Which really kills the point I think?

File: client/my-sites/plans-grid/components/comparison-grid/index.tsx
Comment (zaguiini) [reply client/my-sites/plans-grid/components/comparison-grid/index.tsx:462]: I can't really see how it'd be useful. The function will be redefined anyway and its cost is marginal. Rendering will happen either way as soon as something else in the tree changes, so I don't think we should bother that much about it.

**Comment by zaguiini** [comment]
I'll merge this tomorrow along with deploying a new version of ETK and Happyblocks. Is it necessary, though? I think so for the sake of keeping the code up to date, but it's not really changing anything on these two plugins. @Automattic/martech could you please clarify that to me?

File: client/my-sites/plans-grid/components/actions.tsx
Comment (niranjan-uma-shankar) [inline client/my-sites/plans-grid/components/actions.tsx:126]: Looks like we're introducing a new string. These either need to have translations completed before deploy, or the change should be hidden for non-EN users.

Screenshot 2023-10-10 at 1 06 57 PM

**Review by niranjan-uma-shankar — DISMISSED** [review]
This change doesn't seem to work for me when I'm logged out. Here are the steps I tried:

* In an incognito window, I navigated to https://container-serene-jepsen.calypso.live/setup/new-hosted-site/?flags=plans/hosting-trial.
* I got redirected to https://container-serene-jepsen.calypso.live/start/hosting/user-hosting
* After signing with an email, I got [this step](https://d.pr/i/q0r0rX).
* I selected "Build a new site from scratch", and then selected the data centre in the next step. The next step which is the plans step looked like this:

Screenshot 2023-10-10 at 12 59 32 PM

Also noting that when I tried booting from calypso.localhost with `ENABLE_FEATURES=plans/hosting-trial yarn start`, the plans step looked like this:

Screenshot 2023-10-10 at 1 01 48 PM

Noting that I can see the "Try with Free" button when I'm logged in and enter the setup flow with the calypso.live link, but I couldn't see it with calypso.localhost (got the same screen I shared above with the Business Trial).

Screenshot 2023-10-10 at 1 04 05 PM

* When I clicked on the "Try with Free", I was on the `/setup/transferring-hosted-site/processing?siteId=224535522` page for several minutes. Not sure if I'm doing something wrong here.

**Comment by zaguiini** [comment]
Thanks for the review, @niranjan-uma-shankar

>This change doesn't seem to work for me when I'm logged out

It's not supposed to. The `/setup/new-hosted-site` flow is gated for logged-in users.

>After signing with an email, I got [this step]().
I selected "Build a new site from scratch", and then selected the data centre in the next step. The next step which is the plans step looked like this:

That's because when you login through an email link you're redirected to wordpress.com. The code for this branch is not there yet, and the feature flag is disabled.

>Also noting that when I tried booting from calypso.localhost with ENABLE_FEATURES=plans/hosting-trial yarn start, the plans step looked like this:

My only guess is that the code is not checked out locally. The old version (what you see in the screen shot, a separate plan), is essentially what this PR changes, so it doesn't make sense not to have it unless you're not in sync with remote and in the correct branch 🤔 

>When I clicked on the "Try with Free", I was on the /setup/transferring-hosted-site/processing?siteId=224535522 page for several minutes. Not sure if I'm doing something wrong here.

No, you're correct. The free trial full flow is not yet working 100%. We have an issue to queue the Atomic transfer and then redirect to `/home/%s`.

File: client/my-sites/plans-grid/components/actions.tsx
Comment (zaguiini) [reply client/my-sites/plans-grid/components/actions.tsx:126]: We don't need to rush. The feature is behind a flag and the release date is scheduled for a month from now. We are going to send stuff for translations, but it will be done in a follow-up.

**Comment by zaguiini** [comment]
This is going to be put on hold since we are going to implement some design changes. cc @vinimotaa

File: client/my-sites/plans-grid/components/actions.tsx
Comment (niranjan-uma-shankar) [reply client/my-sites/plans-grid/components/actions.tsx:126]: Sounds good, thanks for clarifying.

**Comment by niranjan-uma-shankar** [comment]
Thanks for clarifying @zaguiini . Regarding your other question:

> I think so for the sake of keeping the code up to date, but it's not really changing anything on these two plugins. https://github.com/orgs/Automattic/teams/martech could you please clarify that to me?

I'm not sure who in MarTech would know about this, maybe @chriskmnds ? AFAIK we don't work closely with these plugins.

**Comment by p-jackson** [comment]
@zaguiini let's go with this design: pdKhl6-2Hm-p2#comment-4423

I really appreciate all the expert reviews 👍

**Comment by zaguiini** [comment]
@Automattic/quake please review and change this PR as needed 👍 🚀

File: client/my-sites/plans-grid/components/billing-timeframe.tsx
Comment (chriskmnds) [inline client/my-sites/plans-grid/components/billing-timeframe.tsx]: I need to have a look here. `flowname` shouldn't leak further into plans-grid as that's a concept we are removing.

~Please wait for explicit approval from code owners before merging anything with this PR (and ping if we are not timely).~

**Review by chriskmnds — DISMISSED** [review]

File: client/my-sites/plans-grid/components/billing-timeframe.tsx
Comment (zaguiini) [reply client/my-sites/plans-grid/components/billing-timeframe.tsx]: It's OK, we can move to `intent`. It's just a bad habit that dies hard :)

**Review by p-jackson — COMMENTED** [review]
I leave the code review to the experts but I've given this a test and everything's looking good.

Desktop/Mobile layout looking good, and existing CTAs still work as expected.

File: client/my-sites/plans-grid/components/billing-timeframe.tsx
Comment (chriskmnds) [reply client/my-sites/plans-grid/components/billing-timeframe.tsx]: Thanks for giving me a chance to respond 😬 

To me, both `flowName` and `intent` bring similar context into the `plans-grid`. `intent` is stronger in that it can apply to both admin and signup (as the case with Newsletter), and having a single concept throughout is likely to lead to less complexity. However, it's still context that we ought to try and minimize its use/effect, or perhaps eliminate altogether from `plans-grid` UI components. 

We've been trying to separate data concerns from UI and that's something we ought to keep doing. I think this case may suffice for a different control/design with that in mind. 

So instead of putting together the refund period here (it very much sounds like "pricing meta" concern to me) and conditioning with:

```
	if ( ! isAnyHostingFlow( flowName ) || isFreePlan( planSlug ) ) {
		return null;
	}
```

I would propose the following:

1. Compute the refund period in `usePricingMetaForGridPlans` and include as a property of `PricingMeta` e.g. `refundPeriod?: number`. right [here](https://github.com/Automattic/wp-calypso/blob/trunk/client/my-sites/plans-features-main/hooks/data-store/use-pricing-meta-for-grid-plans.ts#L207-L214), so something like:

```
{
    originalPrice: planPrices[ planSlug ]?.originalPrice,
    discountedPrice: planPrices[ planSlug ]?.discountedPrice,
    billingPeriod: pricedAPIPlan?.bill_period,
    currencyCode: pricedAPIPlan?.currency_code,
    introOffer: sitePlan?.introOffer,
    expiry: sitePlan?.expiry,
    ...( pricedAPIPlan?.bill_period && ! isFreePlan( planSlug ) && { refundPeriod: pricedAPIPlan?.bill_period === 365 : 14 : 7 } )
}
```

2. Add an optional property to [PlansGridProps](https://github.com/Automattic/wp-calypso/blob/trunk/client/my-sites/plans-grid/index.tsx#L31) to show/hide the refund notice e.g. `showRefundNoticeInBillingDescriptions` (or something less verbose?), which can condition on the flow or intent that we want the notice to be shown. and from `plans-features-main` (no need for `isFreePlan` check):

```
showRefundNoticeInBillingDescriptions={ isAnyHostingFlow( flowName ) }
```

I think that would be a good separation of concerns. Let me know what you think (does it resonate?) 🙌 
I can also do it in a follow-up if you want to ship earlier.

**Review by chriskmnds — COMMENTED** [review]
Thanks for your patience! I left a comment, which I think merits a bit of a refactor. However, I'd be more than happy to also do it in a follow-up myself (but let's ship this sooner in that case before I move on to something else or we build more around this) 😬 

> I leave the code review to the experts but I've given this a test and everything's looking good.

cmon! I'm definitely less of an expert I think 😆 Just trying to align stuff with the direction in the NPM migration thread (and learn/adjust as we go)

File: client/my-sites/plans-grid/components/billing-timeframe.tsx
Comment (chriskmnds) [reply client/my-sites/plans-grid/components/billing-timeframe.tsx]: > Let me know what you think

Double this! 🙇 I could be wrong, always. Just feels to me there are multiple concerns

**Comment by p-jackson** [comment]
> cmon! I'm definitely less of an expert I think 😆 Just trying to align stuff with the direction in the NPM migration thread (and learn/adjust as we go)

true true 😅 tbh I did have a look at the code and it works for us, but I think you've got a good vision for what a packagized plans grid should look like, so wanted to defer.

Those refactor to add refund period to the plan meta and "show refund notice" as a prop sound good. I'll make that change and then I think we're good to merge?

> I can also do it in a follow-up if you want to ship earlier.

I'll merge but feel free to make further necessary refactors to ensure the plans grid stays reusable.

**Comment by chriskmnds** [comment]
> > cmon! I'm definitely less of an expert I think 😆 Just trying to align stuff with the direction in the NPM migration thread (and learn/adjust as we go)
> 
> true true 😅 tbh I did have a look at the code and it works for us, but I think you've got a good vision for what a packagized plans grid should look like, so wanted to defer.
> 
> Those refactor to add refund period to the plan meta and "show refund notice" as a prop sound good. I'll make that change and then I think we're good to merge?
> 
> > I can also do it in a follow-up if you want to ship earlier.
> 
> I'll merge but feel free to make further necessary refactors to ensure the plans grid stays reusable.

Thanks, and totally. Feel free to merge as-is at any time you are happy. I can follow up from here with the refactor 👍

**Comment by p-jackson** [comment]
@zaguiini I've fixed the monster refactor and lifted the call to `isAnyHostingFlow` up and out of the plans grid.

I can't quite figure out the correct way to use the `usePricingMetaForGridPlans` hook so I've left that.

A bit late now for me to merge.

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/9501765

Some locales (Japanese) have been temporarily machine-translated due to translator availability. All other translations are usually ready within a few days. Untranslated and machine-translated strings will be sent for translation next Monday and are expected to be completed by the following Friday.

Thank you @zaguiini for including a screenshot in the description! This is really helpful for our translators.

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/9501765).

File: client/my-sites/plans-grid/components/actions.tsx
Comment (ddc22) [inline client/my-sites/plans-grid/components/actions.tsx:417]: @zaguiini this part of the code confused me a bit. Given that isFreeTrialPlan is derived by the variable `freeTrialPlanSlug `, Why do we need to pass it in as a parameter to this function ?
Can't we achieve the same thing with the params already available here and possibly a unique parameter identifying the button?  

Moreover, IMO this logic should not live inside the plans grid.

Essentially one can provide the plans grid a set of data to display and then provide a callback to be called on a **plan** button click. If required, a parameters can be returned identifying the **plan** button that was clicked i.e. planSlug. Which we already do if you look closer at `PlansFeaturesMain`

Passing and computing parameters of external domain logic would bloat the code of the plans grid making it hard to manage as a reusable component. 

Tracks events also can be generally triggered outside the  grid based on this approach.
https://github.com/Automattic/wp-calypso/blob/30a9ac440933001a289ed7a5fff49f7d0ecfa803/client/my-sites/plans-features-main/index.tsx#L344.

But open to discuss better way to do this as we gradually evolve the plans grid, so do let me know your thoughts :)

File: client/my-sites/plans-grid/components/billing-timeframe.tsx
Comment (chriskmnds) [inline client/my-sites/plans-grid/components/billing-timeframe.tsx:312]: @zaguiini / @Automattic/quake can you please help me confirm as of when we stopped showing the refund notice in the billing description? Is there any discussion/P2 I can follow on the decision to stop showing it under `/setup/new-hosted-site`? 🙇

Comment (p-jackson) [reply client/my-sites/plans-grid/components/billing-timeframe.tsx:312]: I'm surprised, as far as I know the refund notice should still be there for the hosting flow. I hadn't noticed it was gone!

The `showRefundPeriod` prop is [still being set here for our flow](https://github.com/Automattic/wp-calypso/blob/bf76b613c6419a3114a38bdd6dc9a29b247e7419/client/my-sites/plans-features-main/index.tsx#L854).

[But I see `<FeaturesGrid>` isn't anything with this prop](https://github.com/Automattic/wp-calypso/blob/bf76b613c6419a3114a38bdd6dc9a29b247e7419/packages/plans-grid-next/src/components/features-grid/index.tsx#L219). Is this a regression introduced during the refactor to `@automattic/plans-grid-next`?

Comment (chriskmnds) [reply client/my-sites/plans-grid/components/billing-timeframe.tsx:312]: Interesting. It sounds like a regression then. I'll create an issue and perhaps @southp can coordinate with fixing? Otherwise, I'll get to it when back.

Thanks for confirming @p-jackson 🙇

Comment (chriskmnds) [reply client/my-sites/plans-grid/components/billing-timeframe.tsx:312]: Created issue here https://github.com/Automattic/wp-calypso/issues/89945

@southp can you coordinate with investigating? It's likely that we broke this in the refactors.

Comment (jeyip) [reply client/my-sites/plans-grid/components/billing-timeframe.tsx:312]: https://github.com/Automattic/wp-calypso/pull/90056

**Pull request merged**
